File: tests/api/TextToMath.js
/* global require, describe, it */

var expect = require('expect.js')
var request = require('supertest')

describe('TextToMath Api', function() {

    describe('Calculate', function() {

        it('should return numeric result', function(done) {

            request('localhost')
                .post('/api/calculate')
                .send({ input: '1+1' })
                .expect(200)
                .end(function(err, res) {
                    if (err) {
                        throw err
                    }

                    expect(res.body).to.eql({
                        vars: {},
                        funcs: {},
                        bound: {},
                        result: 2.0
                    })

                    done()
                })
        })

        it('should return an error', function(done) {

            request('localhost')
                .post('/api/calculate')
                .send({ input: '1 = 1'})
                .expect(400)
                .end(function(err, res) {
                    if (err) {
                        throw err
                    }

                    // Error at equals sign
                    expect(res.body).to.eql({
                        error: 'Invalid input: at position 3'
                    })

                    done()
                })
        })

        it('should return a variable', function(done) {

            request('localhost')
                .post('/api/calculate')
                .send({ input: 'a = 1'})
                .expect(200)
                .end(function(err, res) {
                    if (err) {
                        throw err
                    }

                    expect(res.body).to.eql({
                        vars: { a: '1.0' },
                        funcs: {},
                        bound: {},
                        result: 1.0
                    })

                    done()
                })
        })

        it('should return a function', function(done) {

            request('localhost')
                .post('/api/calculate')
                .send({ input: 'a(x) = x + 1' })
                .expect(200)
                .end(function(err, res) {
                    if (err) {
                        throw err
                    }

                    expect(res.body).to.eql({
                        vars: {},
                        funcs: { a: '(x) = (1.0 + x)' },
                        bound: {},
                        result: 0.0
                    })

                    done()
                })
        })

        it('should return a bound variable', function(done) {

            request('localhost')
                .post('/api/calculate')
                .send({ input: 'a := 3' })
                .expect(200)
                .end(function(err, res) {
                    if (err) {
                        throw err
                    }

                    expect(res.body).to.eql({
                        result: 3.0,
                        vars: {},
                        funcs: {},
                        bound: {
                            a: {
                                expr: '3.0',
                                value: 3.0
                            }
                        }
                    })

                    done()
                })
        })

    })

    // Test server data
    // User name: testUser
    // Vars:      a = 2.0
    // Functions: a(x) = 2.0 + x
    // Bound:     y := a
    //
    // User name: testUser2
    // Vars:      b = 3.0
    // Functions: b(x) = 3.0 + x
    // Bound:     z := b
    describe('UserInfo', function() {

        it('should return userInfo', function(done) {

            request('localhost')
                .get('/api/userInfo')
                .set('Cookie', 'user-id=testUser')
                .expect(200)
                .end(function(err, res) {
                    if (err) {
                        throw err
                    }

                    expect(res.body).to.eql({
                        vars: { a: '2.0' },
                        funcs: { a: '(x) = (2.0 + x)' },
                        bound: {
                            y: {
                                expr: 'a',
                                value: 2.0
                            }
                        }
                    })

                    done()
                })
        })

        it('should delete userInfo', function(done) {

            request('localhost')
                .del('/api/userInfo')
                .set('Cookie', 'user-id=testUser')
                .expect(204)
                .end(function(err) {
                    if (err) {
                        throw err
                    }

                    done()
                })
        })

        it('should reject content type', function(done) {

            request('localhost')
                .post('/api/userInfo')
                .set('Cookie', 'user-id=testUser2')
                .set('Content-Type', 'text/plain')
                .send('[{ "op": "remove", "path": "/var/b" }]')
                .expect(415)
                .end(function(err, res) {
                    if (err) {
                        throw err
                    }

                    expect(res.text).to.be('Content type must be application/json-patch+json')

                    done()
                })
        })

        it('should only permit remove', function(done) {

            request('localhost')
                .post('/api/userInfo')
                .set('Cookie', 'user-id=testUser2')
                .set('Content-Type', 'application/json-patch+json')
                .send('[{ "op": "add", "path": "/var/b" }]')
                .expect(400)
                .end(function(err, res) {
                    if (err) {
                        throw err
                    }

                    expect(res.text).to.be('Unpermitted operation')

                    done()
                })
        })

        it('should be unable to apply the patch', function(done) {

            request('localhost')
                .post('/api/userInfo')
                .set('Cookie', 'user-id=testUser2')
                .set('Content-Type', 'application/json-patch+json')
                .send('[{ "op": "remove", "path": "/var/nonexistantVar" }]')
                .expect(400)
                .end(function(err, res) {
                    if (err) {
                        throw err
                    }

                    expect(res.text).to.be('Unable to apply patch')

                    done()
                })
        })

        it('should delete the saved information', function(done) {

            var operations = [
                { op: 'remove', path: '/vars/b' },
                { op: 'remove', path: '/funcs/b' },
                { op: 'remove', path: '/bound/z' }
            ]

            request('localhost')
                .post('/api/userInfo')
                .set('Cookie', 'user-id=testUser2')
                .set('Content-Type', 'application/json-patch+json')
                .send(JSON.stringify(operations))
                .expect(200)
                .end(function(err) {
                    if (err) {
                        throw err
                    }

                    request('localhost')
                        .get('/api/userInfo')
                        .set('Cookie', 'user-id=testUser2')
                        .expect(200)
                        .end(function(err, res) {
                            if (err) {
                                throw err
                            }

                            expect(res.body).to.eql({
                                vars: {},
                                funcs: {},
                                bound: {},
                            })

                            done()
                        })
                })
        })
    })
})

File: api/TextToMath.hs
{-# LANGUAGE OverloadedStrings, DoAndIfThenElse #-}

module TextToMath (
    app
) where

import Calculator
import Calculator.Data.Env
import Calculator.Data.AST
import Calculator.Functions (Function, showDeclaration)
import Control.Applicative (optional, (<$>))
import Control.DeepSeq (($!!))
import Control.Monad (msum)
import Control.Monad.IO.Class (liftIO)
import Data.Acid (AcidState)
import Data.Acid.Advanced (query', update')
import Data.Aeson ((.=))
import Data.Char (isSpace)
import Data.List (isPrefixOf, stripPrefix)
import Data.Map (Map)
import Data.Maybe (fromMaybe)
import Calculator.Data.Decimal
import Happstack.Server hiding (body, result)
import Serializer()
import System.UUID.V4 (uuid)
import UserState
import qualified Control.Exception.Lifted as CEL
import qualified Data.Aeson as Aeson (encode, decode)
import qualified Data.Aeson.Types as Aeson hiding (Result)
import qualified Data.ByteString.Char8 as Char8 (unpack)
import qualified Data.Map as Map

app :: AcidState UserDb -> ServerPart Response
app acid = msum
  [ dir "calculate" (calc acid)
  , dir "userInfo" (getUserInfo acid)       -- GET
  , dir "userInfo" (resetUserInfo acid)     -- DELETE
  , dir "userInfo" (modifyUserInfo acid)    -- POST (patch)
  ]

calc :: AcidState UserDb -> ServerPart Response
calc acid = do
        method POST
        rq <- askRq
        if contentType "application/json" rq then do
            maybeBody <- takeRequestBody rq
            case Aeson.decode $ unBody $ fromMaybe (Body "") maybeBody :: Maybe (Map String String) of
                Just jsonContent -> do
                    let input = fromMaybe "" $ Map.lookup "input" jsonContent
                    userId <- getUserId
                    User variables functions <- query' acid (UserState.GetUser userId)
                    result <- liftIO $ getReturnText input $ Env variables (Map.map fst functions)
                    addCookie Session $ mkCookie "user-id" userId
                    case result of
                        Left err -> do
                            let res = jsonResponse [ "error" .= err ]
                            badRequest res
                        Right ans -> do let newVars = vars ans
                                        let newFuncs = addFunctionText input $ copyFunctionText (funcs ans) functions
                                        let addedFuncs = Map.differenceWith (\a b -> if a /= b then Just a else Nothing) newFuncs functions
                                        update' acid (UserState.SetUser userId $ User newVars newFuncs)
                                        let res = jsonResponse $ makeJSON (Map.mapWithKey (varsToJSON $ boundResults ans) $ vars ans)
                                                                          (Map.mapWithKey funcsToJSON addedFuncs)
                                                                          ("result" .= answer ans)
                                        ok res 
                Nothing -> badRequest $ toResponse ("Unable to parse body" :: String)
        else resp 415 $ toResponse ("Content type must be application/json" :: String)
        where -- Extract the user-entered body of the function
              getFunctionText input = dropWhile isSpace $ drop 1 $ dropWhile (not . (==) '=') input
              -- Add the correct function text for new or modified functions
              addFunctionText input = Map.map (\(f,name) -> if null name then (f, getFunctionText input) else (f,name))
              makeJSON vs fs res = [ "vars" .= vs
                                   , "funcs" .= fs
                                   , res
                                   ]
              -- Copy the text of the original functions to the new
              -- functions, with a default value of "".
              -- Union works since the user can't delete functions here
              copyFunctionText newFuncs = Map.unionWith (\new@(newF,_) (oldF, text) -> if newF == oldF then (newF,text) else new) (Map.map (\x -> (x,"")) newFuncs)

getUserInfo :: AcidState UserDb -> ServerPart Response
getUserInfo acid = do
    method GET
    userId <- getUserId
    User variables functions <- query' acid (UserState.GetUser userId)
    addCookie Session $ mkCookie "user-id" userId

    -- no op calculation to force bound vars to get calculed
    result <- liftIO $ getReturnText "0" (Env variables (Map.map fst functions))
    case result of
        Left err -> ok $ jsonResponse [ "error" .= err ]
        Right ans -> ok $ jsonResponse [ "vars" .= Map.mapWithKey (varsToJSON (boundResults ans)) (vars ans)
                                       , "funcs" .= Map.mapWithKey funcsToJSON functions
                                       ]
    
varsToJSON :: Map String Decimal -> String -> AST -> Aeson.Value
varsToJSON _ _ (Number n) = Aeson.object [ "value" .= n ]
varsToJSON bound v expr =  Aeson.object [ "value" .= (bound Map.! v)
                                        , "expr" .= show expr
                                        ]

funcsToJSON :: String -> (Function, String) -> Aeson.Value
funcsToJSON k (f, text) = Aeson.object [ "decl" .= (k ++ showDeclaration f)
                                       , "def" .= text
                                       ]

resetUserInfo :: AcidState UserDb -> ServerPart Response
resetUserInfo acid = do
    method DELETE
    userId <- getUserId
    update' acid $ UserState.SetUser userId UserState.newUser
    noContent $ toResponse ()

modifyUserInfo :: AcidState UserDb -> ServerPart Response
modifyUserInfo acid = do
    method POST
    rq <- askRq
    if contentType "application/json-patch+json" rq then do
        userId <- getUserId
        maybeBody <- takeRequestBody rq
        let b = unBody $ fromMaybe (Body "") maybeBody
        case Aeson.decode b of
            Just values ->
                if all isRemove values then do
                    user <- query' acid $ UserState.GetUser userId
                    let updatedUser = foldl runAction (Just user) values
                    case updatedUser of
                        Just u -> do
                            update' acid $ UserState.SetUser userId u
                            ok $ toResponse ()
                        Nothing ->
                            badRequest $ toResponse ("Unable to apply patch" :: String)
                else
                    badRequest $ toResponse ("Unpermitted operation" :: String)
            Nothing -> badRequest $ toResponse ("Unable to parse body" :: String)
    else resp 415 $ toResponse ("Content type must be application/json-patch+json" :: String)
    where isRemove m = case Map.lookup "op" m of
                            Just o -> o == "remove"
                            Nothing -> False

runAction :: Maybe User -> Map String String -> Maybe User
runAction (Just (User vs fs)) action
    | Map.size action == 2 =
        case Map.lookup "path" action of
            Just p  | "/vars/" `isPrefixOf` p ->
                        let Just varName = stripPrefix "/vars/" p
                        in if Map.member varName vs then
                            Just $ User (Map.delete varName vs) fs
                        else
                            Nothing
                    | "/funcs/" `isPrefixOf` p ->
                        let Just funcName = stripPrefix "/funcs/" p
                        in if Map.member funcName fs then
                            Just $ User vs $ Map.delete funcName fs
                        else
                            Nothing
                    | otherwise -> Nothing
            Nothing -> Nothing
    | otherwise = Nothing
runAction Nothing _ = Nothing

getUserId :: ServerPart String
getUserId = do userId <- optional $ lookCookieValue "user-id"
               case userId of
                   Nothing -> show <$> liftIO uuid
                   Just i -> return i

getReturnText :: String -> Env -> IO (Either String Result)
getReturnText input env = CEL.catch (CEL.evaluate $!! result)
                                    (\e -> return $ Left $ "Invalid input: " ++ show (e :: CEL.ErrorCall))
                                    where result = Right $ calculate input env

jsonResponse :: [Aeson.Pair] -> Response
jsonResponse = addHeader "Content-Type" "application/json" . toResponse . Aeson.encode . Aeson.object

contentType :: String -> Request -> Bool
contentType ct rq = ct `isPrefixOf` rqCt rq
                    where rqCt r = Char8.unpack $ fromMaybe "" $ getHeader ("Content-Type" :: String) r
